Source/Pawnmorphs/Esoteria/UserInterface/Settings/Dialog_BlacklistAnimal.cs

```diff
@@ -0,0 +1,177 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Pawnmorph.Chambers;
+using Pawnmorph.DefExtensions;
+using Pawnmorph.Utilities.Collections;
+using UnityEngine;
+using Verse;
+using Verse.Sound;
+
+namespace Pawnmorph.UserInterface.Settings
+{
+	internal class Dialog_BlacklistAnimal : Window
+	{
+		private static string HEADER_TEXT = "PMBlacklistFormerHumansHeader".Translate();
+		private static string DESCRIPTION_TEXT = "PMBlacklistFormerHumansText".Translate();
+		private static string APPLY_BUTTON_TEXT = "ApplyButtonText".Translate();
+		private static string RESET_BUTTON_TEXT = "ResetButtonText".Translate();
+		private static string CANCEL_BUTTON_TEXT = "CancelButtonText".Translate();
+		private static string FAILED_PARSING_TEXT = "PMFailedToParse".Translate();
+		private static Vector2 APPLY_BUTTON_SIZE = new Vector2(120f, 40f);
+		private static Vector2 RESET_BUTTON_SIZE = new Vector2(120f, 40f);
+		private static Vector2 CANCEL_BUTTON_SIZE = new Vector2(120f, 40f);
+		private static Vector2 RAW_BUTTON_SIZE = new Vector2(28, 28f);
+		private const float SPACER_SIZE = 17f;
+
+		private Dictionary<ThingDef, bool> _blacklistedAnimalDictonary;
+		List<string> _settingsReference;
+		FilterListBox<ThingDef> _animalListBox;
+
+		public Dialog_BlacklistAnimal(List<string> settingsReference)
+		{
+			_settingsReference = settingsReference;
+			_blacklistedAnimalDictonary = new Dictionary<ThingDef, bool>();
+		}
+
+		public override void PostOpen()
+		{
+			base.PostOpen();
+			RefreshList();
+		}
+
+		private void RefreshList()
+		{
+			// Get all animal ThingDefs that are not dryads.
+			IEnumerable<ThingDef> animals = DefDatabase<ThingDef>.AllDefsListForReading.Where(x => x.race?.Animal == true && x.race?.Dryad != true);
+			animals = animals.Except(MorphDef.AllDefs.SelectMany(x => x.AllAssociatedAnimals));
+
+			_blacklistedAnimalDictonary = animals.ToDictionary(x => x, x => _settingsReference.Contains(x.defName) == false);
```

Comment:
The value of NeverFormerHuman from the XML doesn't appear to be read at all when populating this dictionary.  If I'm reading this right, the Apply logic manually sets the value to false for anything not explicitly appearing in the blacklist.  That means opening the dialog, not changing anything, then closing it will end up un-blacklisting any animal set to be blacklisted via XML.  That's almost certainly not what we want.

(I'm not sure we should allow players to un-blacklist animals that are blacklisted via XML anyway, but it definitely shouldn't ignore the XML completely.)

Reply:
You are right.

I had the code there but ended up entirely removing when i was trying to get UI working.

I will need to reconsider how it should work.